File: platform/platform-core/fastify/src/lifecycles/04_context_resolver.lifecycle.mjs
import { readdir } from "fs/promises";
import { existsSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";
import { registerComputeFunctions } from "./04-context-resolver.registry.mjs";

let createRegistry,
  ComputeScope,
  createSdk,
  createSharedContext,
  contextResolverPlugin;
let HAS_RESOLVER = false;

try {
  const resolver = await import("app_yaml_overwrites");
  createRegistry = resolver.createRegistry;
  ComputeScope = resolver.ComputeScope;
  createSdk = resolver.createSdk;
  createSharedContext = resolver.createSharedContext;

  // Attempt to load plugin (checking for both new and legacy names if needed)
  contextResolverPlugin = resolver.configPlugin;

  if (!contextResolverPlugin) {
    // Fallback or explicit import if not exported in main index
    const integrations = await import(
      "app_yaml_overwrites/integrations/fastify"
    );
    contextResolverPlugin =
      integrations.configPlugin || integrations.contextResolverPlugin;
  }

  HAS_RESOLVER = true;
} catch (e) {
  console.warn("[lifecycle:context-resolver] app_yaml_overwrites not available:", e.message);
}

/**
 * Auto-load compute functions from *.compute.mjs files in computed_functions directory.
 *
 * Each file should expose:
 * - register: function - The compute function to register
 * - NAME (optional): string - Name to register under (defaults to filename without .compute.mjs)
 * - SCOPE (optional): ComputeScope - Scope for the function (defaults to ComputeScope.STARTUP)
 *
 * @param {Object} registry - The compute function registry
 * @param {string} [baseDir] - Base directory to scan (defaults to ../computed_functions)
 * @param {Object} [logger] - Logger instance
 * @returns {Promise<string[]>} - List of loaded function names
 */
async function autoLoadComputeFunctions(
  registry,
  baseDir = null,
  logger = console
) {
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = dirname(__filename);

  if (!baseDir) {
    baseDir = join(__dirname, "..", "..", "computed_functions");
  }

  if (!existsSync(baseDir)) {
    return [];
  }

  const loaded = [];

  try {
    const files = await readdir(baseDir);
    const computeFiles = files.filter((f) => f.endsWith(".compute.mjs"));

    // Log files that don't match the *.compute.mjs pattern
    const allMjsFiles = files.filter((f) => f.endsWith(".mjs"));
    const ignoredFiles = allMjsFiles.filter((f) => !f.endsWith(".compute.mjs"));
    if (ignoredFiles.length > 0) {
      logger.info(
        { ignoredFiles },
        "[lifecycle:context-resolver] Ignored files in computed_functions/ (don't match *.compute.mjs pattern)"
      );
    }

    for (const file of computeFiles) {
      const filepath = join(baseDir, file);
      const funcName = file.replace(".compute.mjs", "");

      try {
        // Dynamic import using file:// URL
        const fileUrl = `file://${filepath}`;
        const module = await import(fileUrl);

        // Get required register function
        if (typeof module.register !== "function") {
          logger.warn(
            { filepath },
            `[lifecycle:context-resolver] ${filepath} does not export 'register' function, skipping`
          );
          continue;
        }

        const registerFunc = module.register;

        // Get optional NAME (defaults to filename)
        const name = module.NAME || funcName;

        // Get optional SCOPE (defaults to STARTUP)
        const scope = module.SCOPE || ComputeScope.STARTUP;

        // Register the function
        registry.register(name, registerFunc, scope);
        loaded.push(name);
        logger.debug({ name, scope }, "[lifecycle:context-resolver] Registered compute function");
      } catch (e) {
        logger.error(
          { err: e, filepath },
          `[lifecycle:context-resolver] Error loading compute function from ${filepath}: ${e.message}`
        );
      }
    }
  } catch (e) {
    logger.error(
      { err: e, baseDir },
      `[lifecycle:context-resolver] Error scanning computed_functions directory: ${e.message}`
    );
  }

  return loaded;
}


export async function onStartup(server, config) {
  server.log.info(
    { HAS_RESOLVER },
    "[lifecycle:context-resolver] Initializing context resolver..."
  );

  try {
    if (!HAS_RESOLVER) {
      server.log.warn(
        "[lifecycle:context-resolver] app_yaml_overwrites not installed, context resolver skipping."
      );
      return;
    }

    server.log.info("[lifecycle:context-resolver] Initializing app_yaml_overwrites...");

    // server.config is AppYamlConfig instance (decorated in 01)
    const appConfig = server.config;
    server.log.info(
      { appConfigExists: !!appConfig, appConfigType: typeof appConfig },
      "[lifecycle:context-resolver] Checking server.config"
    );
    if (!appConfig) {
      server.log.warn("[lifecycle:context-resolver] server.config not found. Context resolver skipping.");
      return;
    }

    // Get raw config
    server.log.info("[lifecycle:context-resolver] Extracting raw config from AppYamlConfig");
    const rawConfig = appConfig.toObject
      ? appConfig.toObject()
      : appConfig.getAll
      ? appConfig.getAll()
      : {};

    server.log.info("[lifecycle:context-resolver] Creating compute function registry");
    const registry = createRegistry(server.log);
    registerComputeFunctions(registry, ComputeScope);
    server.log.info("[lifecycle:context-resolver] Built-in compute functions registered");

    // Auto-load compute functions from computed_functions directory
    server.log.info("[lifecycle:context-resolver] Auto-loading compute functions from computed_functions directory");
    const autoLoaded = await autoLoadComputeFunctions(registry, null, server.log);
    if (autoLoaded.length > 0) {
      server.log.info({ autoLoaded }, `[lifecycle:context-resolver] Auto-loaded compute functions: ${autoLoaded.join(", ")}`);
    } else {
      server.log.info("[lifecycle:context-resolver] No additional compute functions found in computed_functions directory");
    }

    // Register external/cloud compute functions from server.externalComputeFunctions
    // These should be set by an earlier lifecycle (e.g., 03-external-compute.lifecycle.mjs)
    const externalComputeFunctions = server.externalComputeFunctions;
    server.log.debug({ hasExternal: !!(externalComputeFunctions && Object.keys(externalComputeFunctions).length) }, "[lifecycle:context-resolver] Checking externalComputeFunctions");
    if (externalComputeFunctions && typeof externalComputeFunctions === "object") {
      const externalLoaded = [];
      for (const [name, funcConfig] of Object.entries(externalComputeFunctions)) {
        try {
          const fn = typeof funcConfig === "function" ? funcConfig : funcConfig?.fn;
          const scope =
            typeof funcConfig === "function"
              ? ComputeScope.STARTUP
              : funcConfig?.scope || ComputeScope.STARTUP;
          if (fn) {
            registry.register(name, fn, scope);
            externalLoaded.push(name);
          }
        } catch (e) {
          server.log.error({ err: e, name }, `[lifecycle:context-resolver] Error registering external compute function '${name}': ${e.message}`);
        }
      }
      if (externalLoaded.length > 0) {
        server.log.info({ externalLoaded }, `[lifecycle:context-resolver] Registered external compute functions: ${externalLoaded.join(", ")}`);
      }
    }

    // Initialize SDK and SharedContext (Standard Compliance)
    // Note: server.sdk might be decorated with AppYamlConfigSDK from 01-app-yaml.mjs,
    // but we need ConfigSDK from app-yaml-overwrites which has getResolved() method.
    // We use a different decorator name 'configSdk' to avoid conflicts.
    let sdkInstance;
    if (createSdk) {
      if (
        server.hasDecorator("configSdk") &&
        typeof server.configSdk?.getResolved === "function"
      ) {
        sdkInstance = server.configSdk;
        server.log.info("[lifecycle:context-resolver] Using existing server.configSdk");
      } else {
        server.log.info("[lifecycle:context-resolver] Creating new ConfigSDK instance");
        sdkInstance = await createSdk({ config: rawConfig, registry: registry });
        if (!server.hasDecorator("configSdk")) {
          server.decorate("configSdk", sdkInstance);
        }
        server.log.info("[lifecycle:context-resolver] Created new ConfigSDK instance");
        server.log.debug(
          { methods: Object.getOwnPropertyNames(Object.getPrototypeOf(sdkInstance)) },
          "[lifecycle:context-resolver] ConfigSDK available methods"
        );
      }
    }

    // SharedContext is now created in 02-create_shared_context.mjs
    // Just verify it exists for the plugin
    if (createSharedContext) {
      const sharedContext = server.sharedContext || createSharedContext();
      if (!server.hasDecorator("sharedContext")) {
        server.decorate("sharedContext", sharedContext);
        server.log.info("[lifecycle:context-resolver] Created fallback SharedContext instance");
      } else {
        server.log.info(
          "[lifecycle:context-resolver] Using existing server.sharedContext from 02-create_shared_context.mjs"
        );
      }
    }

    // Register plugin which handles decoration and STARTUP resolution
    // We pass the manually created SDK to avoid duplication
    try {
      server.log.info("[lifecycle:context-resolver] Registering contextResolverPlugin");
      await server.register(contextResolverPlugin, {
        sdk: sdkInstance, // Pass the manual ConfigSDK instance
        registry: registry,
        decoratorName: "configSdk", // Use 'configSdk' to avoid conflict with AppYamlConfigSDK on 'sdk'
        resolveOnStartup: true,
        perRequestResolution: true, // Enable if needed (FastAPI has manual middleware, Fastify has hook)
        logger: server.log,
      });
      server.log.info("[lifecycle:context-resolver] contextResolverPlugin registered successfully");
    } catch (pluginErr) {
      server.log.error(
        { err: pluginErr },
        `[lifecycle:context-resolver] Failed to register contextResolverPlugin: ${pluginErr.message}`
      );
    }

    // Decorate server with registry for healthz access (always do this, even if plugin failed)
    if (!server.hasDecorator("contextRegistry")) {
      server.decorate("contextRegistry", registry);
      server.log.info("[lifecycle:context-resolver] Decorated server with contextRegistry");
    }
    if (!server.hasDecorator("contextRawConfig")) {
      server.decorate("contextRawConfig", rawConfig);
      server.log.info("[lifecycle:context-resolver] Decorated server with contextRawConfig");
    }

    // Log all registered compute functions
    const allFunctions = registry.list();
    server.log.info(
      { count: allFunctions.length, functions: allFunctions },
      "[lifecycle:context-resolver] app_yaml_overwrites initialized"
    );
  } catch (err) {
    server.log.error({ err, hookName: "04-context-resolver" }, "[lifecycle:context-resolver] Context resolver initialization failed");
    throw err;
  }
}

File: platform/fastapi_server/config/lifecycle/04_context_resolver.lifecycle.py
import logging
import os
import glob
import importlib.util
from pathlib import Path
from datetime import datetime
from fastapi import FastAPI

logger = logging.getLogger("lifecycle:context_resolver")

# Load registry module (numeric prefix prevents standard imports)
_registry_spec = importlib.util.spec_from_file_location(
    "context_resolver_registry",
    Path(__file__).parent / "04_context_resolver_registry.py",
)
_registry_mod = importlib.util.module_from_spec(_registry_spec)
_registry_spec.loader.exec_module(_registry_mod)
register_compute_functions = _registry_mod.register_compute_functions

try:
    from app_yaml_overwrites import create_registry, ComputeScope, create_sdk, create_shared_context
    from app_yaml_overwrites.integrations.fastapi import resolve_startup
    HAS_RESOLVER = True
except ImportError:
    HAS_RESOLVER = False


def auto_load_compute_functions(registry, base_dir=None):
    """
    Auto-load compute functions from *.compute.py files in computed_functions directory.

    Each file should expose:
    - register: callable - The compute function to register
    - NAME (optional): str - Name to register under (defaults to filename without .compute.py)
    - SCOPE (optional): ComputeScope - Scope for the function (defaults to ComputeScope.STARTUP)
    """
    if base_dir is None:
        base_dir = Path(__file__).parent.parent.parent / "computed_functions"
    else:
        base_dir = Path(base_dir)

    if not base_dir.exists():
        logger.debug("computed_functions directory not found: %s", base_dir)
        return []

    loaded = []
    pattern = str(base_dir / "*.compute.py")

    # Log files that don't match the *.compute.py pattern
    all_py_files = list(base_dir.glob("*.py"))
    compute_files = [Path(f) for f in glob.glob(pattern)]
    ignored_files = [f for f in all_py_files if f not in compute_files]
    if ignored_files:
        logger.debug(
            "Ignored files in computed_functions/ (don't match *.compute.py pattern): %s",
            [f.name for f in ignored_files],
        )

    for filepath in glob.glob(pattern):
        filepath = Path(filepath)
        module_name = filepath.stem  # e.g., "my_func.compute"
        func_name = module_name.replace(".compute", "")  # e.g., "my_func"

        try:
            # Dynamic import
            spec = importlib.util.spec_from_file_location(module_name, filepath)
            if spec is None or spec.loader is None:
                logger.warning("Could not load spec for %s", filepath)
                continue

            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)

            # Get required register function
            if not hasattr(module, "register"):
                logger.warning("%s does not export 'register' function, skipping", filepath)
                continue

            register_func = module.register

            # Get optional NAME (defaults to filename)
            name = getattr(module, "NAME", func_name)

            # Get optional SCOPE (defaults to STARTUP)
            scope = getattr(module, "SCOPE", ComputeScope.STARTUP)

            # Register the function
            registry.register(name, register_func, scope)
            loaded.append(name)
            logger.debug("Loaded compute function: %s (scope: %s)", name, scope)

        except Exception as e:
            logger.error("Error loading compute function from %s: %s", filepath, e, exc_info=True)

    return loaded




async def onStartup(app: FastAPI, config: dict):
    """Initialize app_yaml_overwrites on startup."""
    logger.info("Starting context_resolver lifecycle hook...")
    try:
        if not HAS_RESOLVER:
            logger.warning("app_yaml_overwrites not installed, context resolver skipping.")
            return

        logger.info("Initializing app_yaml_overwrites...")

        # Get config from app.state (set by 01_app_yaml)
        app_config = getattr(app.state, "config", None)
        if not app_config:
            logger.warning("app.state.config not found. Context resolver skipping.")
            return

        logger.debug("Retrieved app_config from app.state")

        # Get raw config dictionary
        if hasattr(app_config, "get_all"):
            raw_config = app_config.get_all()
        elif hasattr(app_config, "to_dict"):
            raw_config = app_config.to_dict()
        else:
            raw_config = {}

        registry = create_registry()
        logger.debug("Created compute registry")
        register_compute_functions(registry, ComputeScope)
        logger.debug("Registered built-in compute functions")

        # Auto-load compute functions from computed_functions directory
        auto_loaded = auto_load_compute_functions(registry)
        if auto_loaded:
            logger.info("Auto-loaded compute functions: %s", auto_loaded)
        else:
            logger.debug("No auto-loaded compute functions found")

        # Register external/cloud compute functions from app.state
        # These should be set by an earlier lifecycle (e.g., 02.5_cloud_compute.lifecycle.py)
        external_compute_functions = getattr(app.state, 'external_compute_functions', None)
        if external_compute_functions:
            external_loaded = []
            for name, func_config in external_compute_functions.items():
                try:
                    fn = func_config.get('fn') if isinstance(func_config, dict) else func_config
                    scope = func_config.get('scope', ComputeScope.STARTUP) if isinstance(func_config, dict) else ComputeScope.STARTUP
                    registry.register(name, fn, scope)
                    external_loaded.append(name)
                except Exception as e:
                    logger.error("Error registering external compute function '%s': %s", name, e, exc_info=True)
            if external_loaded:
                logger.info("Registered external compute functions: %s", external_loaded)
        else:
            logger.debug("No external compute functions found on app.state")

        # Store registry on app.state for route access
        app.state.context_registry = registry
        app.state.context_raw_config = raw_config

        # Create SDK (Standard Compliance)
        app.state.sdk = create_sdk({"config": raw_config, "registry": registry})
        logger.debug("Created SDK and stored on app.state")

        # SharedContext is now created in 02_create_shared_context.py
        # Just use existing or create fallback
        shared_context = getattr(app.state, 'sharedContext', None) or create_shared_context()
        if getattr(app.state, 'sharedContext', None) is None:
            app.state.sharedContext = shared_context
            logger.info("Created fallback SharedContext instance")
        else:
            logger.info("Using existing app.state.sharedContext from 02_create_shared_context.py")

        # Resolve STARTUP config and store in app.state.resolved_config
        logger.info("Resolving STARTUP-scoped compute functions...")
        await resolve_startup(
            app=app,
            config=raw_config,
            registry=registry,
            state_property="resolved_config",
            shared_context=shared_context
        )
        logger.debug("STARTUP resolution complete")

        # Log all registered compute functions
        all_functions = registry.list()
        logger.info("app_yaml_overwrites initialized.")
        logger.info("Registered compute functions (%d):", len(all_functions))
        for func_name in all_functions:
            logger.debug("  - %s", func_name)

        logger.info("context_resolver lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("context_resolver lifecycle hook failed: %s", exc, exc_info=True)
        raise
